<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <xacro:macro name="mobile_robot_macro" params="prefix">
    <!-- define the base-link origin to lie at floor level, between the drive wheels-->
    <!--main body is a simple box; origin is a center of box-->
    <xacro:property name="bodylen" value="0.5061"/>
    <xacro:property name="bodywidth" value="0.4072"/>
    <xacro:property name="bodyheight" value="0.16"/>
    <!--clearance from bottom of box to ground-->
    <!-- derived values-->
    <xacro:property name="half_bodylen" value="${bodylen/2.0}"/>
    <xacro:property name="half_bodyheight" value="${bodyheight/2.0}"/>
    <!-- placement of main body relative to base link frame -->
    <xacro:property name="bodyOX" value="${-half_bodylen}"/>
    <xacro:property name="bodyOY" value="0"/>
    <xacro:property name="bodyOZ" value="0.35"/>

    <!-- define the drive-wheel dimensions-->
    <xacro:property name="tirediam" value="0.3302"/>
    <xacro:property name="tirerad" value="${tirediam/2.0}"/>
    <xacro:property name="tirewidth" value="0.06985"/>
    <!-- "track" is the distance between the drive wheels -->
    <xacro:property name="track" value="0.51015"/>

    <xacro:property name="M_PI" value="3.1415926535897931"/>
    <xacro:property name="boschwidth" value="0.0381"/>
    <xacro:property name="casterdrop" value="0.05"/>
    <xacro:property name="frontdrop" value="${tirerad+0.02-0.1}"/>
    <xacro:property name="bracketwidth" value="0.1175"/>
    <xacro:property name="bracketheight" value="0.16"/>
    <xacro:property name="bracketthick" value="0.0508"/>
    <xacro:property name="bracketangle" value="0.7854"/>
    <xacro:property name="casterwidth" value="0.0826"/>
    <xacro:property name="casterdiam" value="0.2286"/>

    <!--model will assign inertia matrix dominated by main body box -->
    <xacro:macro name="default_inertial" params="mass">
      <inertial>
        <mass value="${mass}"/>
        <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
      </inertial>
    </xacro:macro>

    <xacro:macro name="wheel" params="prefix reflect">
      <link name="front_hoop_${prefix}">
        <visual>
          <geometry>
            <box size="${boschwidth} ${boschwidth} ${frontdrop}"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </visual>
        <collision>
          <geometry>
            <box size="${boschwidth} ${boschwidth} ${frontdrop}"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <xacro:default_inertial mass="0.2"/>
      </link>
      <joint name="hoop2base_${prefix}" type="fixed">
        <parent link="base_link"/>
        <child link="front_hoop_${prefix}"/>
        <origin xyz="${bodylen/2+bodyOX-boschwidth/2} ${reflect*bodywidth/2-reflect*boschwidth/2} ${-frontdrop/2-bodyheight/2+bodyOZ}"/>
      </joint>
      <link name="${prefix}_wheel">
        <visual>
          <geometry>
            <cylinder radius="${tirerad}" length="${tirewidth}"/>
          </geometry>
        </visual>
        <collision>
          <geometry>
            <cylinder radius="${tirerad}" length="${tirewidth}"/>
          </geometry>
        </collision>
        <inertial>
          <!--assign inertial properties to drive wheels -->
          <mass value="1.5"/>
          <inertia ixx="0.1" ixy="0" ixz="0" iyy="0.1" iyz="0" izz="0.1"/>
        </inertial>
      </link>
      <joint name="${prefix}_wheel_joint" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="base_link"/>
        <child link="${prefix}_wheel"/>
        <origin xyz="0 ${reflect*track/2} ${tirerad}" rpy="0 ${M_PI/2} ${M_PI/2}"/>
        <limit effort="100" velocity="25"/>
        <joint_properties damping="0.0" friction="0.0"/>
      </joint>
    </xacro:macro>

    <xacro:macro name="caster" params="prefix reflect">
      <link name="castdrop_${prefix}">
        <visual>
          <geometry>
            <box size="${boschwidth} ${boschwidth} ${casterdrop}"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </visual>
        <collision>
          <geometry>
            <box size="${boschwidth} ${boschwidth} ${casterdrop}"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <xacro:default_inertial mass="0.2"/>
      </link>
      <joint name="cast2base_${prefix}" type="fixed">
        <parent link="base_link"/>
        <child link="castdrop_${prefix}"/>
        <origin xyz="${-bodylen/2+bodyOX+boschwidth/2} ${reflect*bodywidth/2-reflect*boschwidth/2} ${-casterdrop/2-bodyheight/2+bodyOZ}"/>
      </joint>
      <link name="brackettop_${prefix}">
        <visual>
          <geometry>
            <box size="${bracketwidth} ${bracketthick} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </visual>
        <collision>
          <geometry>
            <box size="${bracketwidth} ${bracketthick} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="0 0 0"/>
        </collision>
        <xacro:default_inertial mass="0.1"/>
      </link>
      <joint name="cast2bracket_${prefix}" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="castdrop_${prefix}"/>
        <child link="brackettop_${prefix}"/>
        <origin xyz="0 0 ${-casterdrop/2}" rpy="0 0 ${M_PI/2}"/>
        <joint_properties damping="0.0" friction="0.0"/>
      </joint>
      <link name="bracketside1_${prefix}">
        <visual>
          <geometry>
            <box size="${bracketthick} ${bracketheight} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="${M_PI/2} ${-bracketangle} ${M_PI/2}"/>
        </visual>
        <collision>
          <geometry>
            <box size="${bracketthick} ${bracketheight} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="${M_PI/2} ${-bracketangle} ${M_PI/2}"/>
        </collision>
        <xacro:default_inertial mass="0.05"/>
      </link>
      <joint name="brack2top1_${prefix}" type="fixed">
        <parent link="brackettop_${prefix}"/>
        <child link="bracketside1_${prefix}"/>
        <origin xyz="${bracketwidth/2} .04 -${bracketheight/2-.02}" rpy="0 0 0"/>
      </joint>
      <link name="bracketside2_${prefix}">
        <visual>
          <geometry>
            <box size="${bracketthick} ${bracketheight} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="${M_PI/2} ${-bracketangle} ${M_PI/2}"/>
        </visual>
        <collision>
          <geometry>
            <box size="${bracketthick} ${bracketheight} .005"/>
          </geometry>
          <origin xyz="0 0 0" rpy="${M_PI/2} ${-bracketangle} ${M_PI/2}"/>
        </collision>
        <xacro:default_inertial mass="0.05"/>
      </link>
      <joint name="brack2top2_${prefix}" type="fixed">
        <parent link="brackettop_${prefix}"/>
        <child link="bracketside2_${prefix}"/>
        <origin xyz="${-bracketwidth/2} .04 -${bracketheight/2-.02}" rpy="0 0 0"/>
      </joint>
      <link name="${prefix}_casterwheel">
        <visual>
          <geometry>
            <cylinder radius="${casterdiam/2}" length="${casterwidth}"/>
          </geometry>
        </visual>
        <collision>
          <geometry>
            <cylinder radius="${casterdiam/2}" length="${casterwidth}"/>
          </geometry>
        </collision>
        <!-- accept default inertial properties for caster wheels-->
        <xacro:default_inertial mass="0.3"/>
      </link>
      <joint name="${prefix}_caster_joint" type="continuous">
        <axis xyz="0 0 1"/>
        <parent link="bracketside1_${prefix}"/>
        <child link="${prefix}_casterwheel"/>
        <origin xyz="${-casterwidth/2-.02} .053 -.053" rpy="0 ${M_PI/2} 0"/>
        <limit effort="100" velocity="25"/>
        <joint_properties damping="0.0" friction="0.0"/>
      </joint>
    </xacro:macro>

    <link name="robot_body">
      <visual>
        <geometry>
          <box size="${bodylen} ${bodywidth} ${bodyheight}"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
      </visual>
      <collision>
        <geometry>
          <box size="${bodylen} ${bodywidth} ${bodyheight}"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
      </collision>
      <xacro:default_inertial mass="100"/>
    </link>

    <!-- Base link is the footprint of the robot -->
    <link name="base_link"/>

    <joint name="body_link" type="fixed">
      <origin xyz="${bodyOX} ${bodyOY} ${bodyOZ}" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="robot_body"/>
    </joint>

    <!-- <xacro:caster prefix="left" reflect="1"/> -->
    <!-- <xacro:caster prefix="right" reflect="-1"/> -->
    <xacro:wheel prefix="left" reflect="1"/>
    <xacro:wheel prefix="right" reflect="-1"/>
    <xacro:caster prefix="rear" reflect="0"/>
    <gazebo reference="right_wheel">
      <mu1>100000.0</mu1>
      <mu2>100000.0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>
    <gazebo reference="left_wheel">
      <mu1>100000.0</mu1>
      <mu2>100000.0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="rear_casterwheel">
      <mu1>1000.0</mu1>
      <mu2>1000.0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>
    <gazebo>
      <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
        <jointName>cast2bracket_rear, rear_caster_joint</jointName>
        <updateRate>60</updateRate>
      </plugin>
    </gazebo>
<!-- 
     <gazebo reference="right_casterwheel">
      <mu1>1000.0</mu1>
      <mu2>1000.0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>


    <gazebo reference="left_casterwheel">
      <mu1>1000.0</mu1>
      <mu2>1000.0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>
    
    <gazebo>
      <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
        <jointName>cast2bracket_right, cast2bracket_left, right_caster_joint, left_caster_joint</jointName>
        <updateRate>60</updateRate>
      </plugin>
    </gazebo> -->
    <!-- publish joint state for passive wheels -->
  </xacro:macro>
</robot>
